### DISCUSSION TITLE
[Early Feedback] External Style Import Unification

### DISCUSSION THREAD

**Discussion by FredKSchott** [withastro/roadmap#28] [Proposal]
There are currently ~three~ four different ways to bring an external CSS file onto the page:

1. `<link href={Astro.resolve("./X.css")} />`
2. ~~`<link rel="stylesheet" href={xUrl}> // import xUrl from './X.css?url';`~~ Update: Not supported by Vite
3. `<style>@import './X.css';</style>`
4. `import './X.css'; // in Astro component frontmatter`

It's difficult for us as maintainers to document, test and support so many different way to do things. Each method may support different features, or imply some different behavior. It's equally difficult for users to understand which is the "right" way to do things. 

Now that #12 and #26 exist, I think we can simplify this somewhat. Here's my proposal, which is following the same logic of #12 and #26 around `<head>` behavior and not adding magic (like bundling) to existing HTML elements (like `<link>`) without user opt-in.

#### 1. ✅ `<link href="/global.css" />` or `<link href="https://..." />`

- **Usecase:** Referencing a `/public` or external CSS file. 
- **Usecase:** I need complete control over this `<link>` tag in the browser.
- No longer supported to use with `Astro.resolve()`! (see other options below)
- Not touched by Astro compiler, runtime, or Vite. Will be left as-is in the final output. 
- Must point to a file in `public/` or an external CSS URL. This will not be bundled/optimized.
  - If you need to point to a specific CSS file in `src/`, see option 2.
- Must exist in `<head>`

#### 2. ❌ `<link rel="stylesheet" href={xUrl}> // import xUrl from './X.css?url';`

- Vite doesn't really support this import. @matthewp fixed a bug where this now returns the correct URL for `xUrl`, but it still won't bundle this CSS file as you would expect. This means that the resolved URL might not actually exist in the final build, if that file had been bundled. 
- It would be great to support this, but right now this is considered a Vite limitation. @matthewp to confirm for the final RFC.
- reproduction: https://stackblitz.com/edit/vite-1jaulg?file=dist%2Fassets%2Findex.dd7a6b16.js

#### 3. ✅ `<style @component>@import './X.css';</style>`

_Note: I'm using `@component` here in explicit reference to #12, but if that is unfamiliar then ignore it for now. This is essentially the same as our current v0.20 default scoping and bundling behavior of Astro component `<style>`._

- **Usecase:** Importing a CSS file that you want optimized with the rest of your site.
- Imported CSS is bundled with the rest of the Astro component-level CSS on the page
- Imported CSS is not scoped. Only the CSS defined inline inside the `<style @component>` element is scoped.
- Resolved by Vite, which means that `@import './X.css';` will inherit from all Vite support, including aliases and npm package `@import` support.

#### 4. ✅ `import './X.css'; // in Astro component frontmatter`

- Supported, but we would not really document this and instead recommend 3. in our documentation and examples.
- Same build/bundle behavior as if you had done 3.
- In a previous draft, the proposal was to completely disable this. However, we wouldn't be able to disable you importing a JS file that then did it's own CSS file import, so we decided to just leave the support in for those who ended up using it.

## Feedback?

**Comment by jonathantneal** [comment]
For HTML `<link>`, I am hoping that https://github.com/withastro/rfcs/discussions/23 will cover resolution, and that https://github.com/withastro/rfcs/discussions/15 will cover hoisting.

For CSS `@import` within `<style>`, I am hoping this can be handled consistently with `import` within `<script>`. The resolution proposal mentions a custom `local:` scheme (like `npm:` but for local references) as an alternative to a custom attribute, and it seems like that solution could potentially cover this. If so, it would be cool to have one consistent way to reference things in so many places.

**Comment by IanVS** [comment]
Another possible way to bring in css is from an `import` in a client-side script.  That was working for me in 0.20, and still works in 0.21 in dev, but not when built.  I'm guessing it may not be something you want to support, but maybe it's worth calling out explicitly whether it's supported or not.

**Comment by matthewp** [comment]
There's a 4th way:

```astro
---
import xUrl './X.css?url';
---
<link rel="stylesheet" href={xUrl}>
```

**Reply by FredKSchott** [reply]
yes! thanks, adding to the original post now

**Reply by FredKSchott** [reply]
oh weird, this actually doesn't work. In the final build you get `xUrl` as `i`, where `var i='export default "data:text/css;base64,I2FwcCB7CiAgZm9udC1mYW1pbHk6IEF2ZW5pciwgSGVsdmV0aWNhLCBBcmlhbCwgc2Fucy1zZXJpZjsKICAtd2Via2l0LWZvbnQtc21vb3RoaW5nOiBhbnRpYWxpYXNlZDsKICAtbW96LW9zeC1mb250LXNtb290aGluZzogZ3JheXNjYWxlOwogIHRleHQtYWxpZ246IGNlbnRlcjsKICBjb2xvcjogIzJjM2U1MDsKICBtYXJnaW4tdG9wOiA2MHB4Owp9Cg=="';`

https://stackblitz.com/edit/vite-1jaulg?file=dist/assets/index.8468bb85.js

You see something similarly weird in dev as well.

Update: I think this is a bug with Vite and using a CSS ESM import + `?url`, I can't reproduce with any other asset type. I'll still update the original post to include this.

**Reply by matthewp** [reply]
I think we should consider this a bug.

**Reply by FredKSchott** [reply]
agreed. OP updated!

**Comment by innermatrix** [comment]
@FredKSchott One thorn in my side right now is that I don't see any good documentation in 0.21, and some inconsistent behavior, with regards to

1. whether snowpack aliases should work with CSS imports (and if so, in which contexts) and 
2. more broadly, the intended best practice around including CSS from 3rd party libraries

The main concern that I have with the proposal as given here is that form # 4, which you are proposing to eliminate, is the only form that I believe works with snowpack aliases today. As a result, eliminating that form without providing a replacement, it looks like we would complicate/uglify inclusion of library CSS?

I would like to see this addressed at some point, and this RFC seems like it might be a good place to fold that in — but I defer to you wrt the scope of this discussion.

**Reply by FredKSchott** [reply]
Looks like that link is to an unrelated discussion, mind updating so that I can understand better?

Agreed that docs need a revamp, I'll add that to the #1 document to track for now. 

Inconsistent behavior is exactly what we're trying to solve for with proposals like this one, #26, and #12 so I definitely feel your pain there. I think it would be safe to assume that whatever we standardize on here would support aliases, and I think that that would make sense as an explicit goal in the final RFC for this proposal. The fact that some methods do today and some don't is one of the main reasons to tackle this work now!

**Reply by innermatrix** [reply]
Sorry, that link wasn't supposed to be a link, it was a reference to the form # 4 in your list above

**Reply by FredKSchott** [reply]
Ah, got it. Yup, I'll add a mention that the new version of 3. would support everything that an ESM import already supports (including aliases). The idea is to pass more of this logic through Vite, so that we don't have to create a second implementation for things like aliases. 

fwiw 2. would also support this, but I agree that that's less useful if you're just working on an individual component.

**Reply by innermatrix** [reply]
K; in that case, I have a related question: how does any of this interact with scss? In 0.20 (which I am stuck on for unrelated reasons) import aliases don't work in `@use` statements in SCSS — would the revised form 3 make aliases available in those as well?

**Reply by FredKSchott** [reply]
I think that that would be considered unrelated to this discussion, since that's really an issue of how Sass + Vite work together. I'm not sure what the answer is but you might have some luck googling around for phrases like "vite sass alias @use" and finding an answer in the larger Vite community.

**Reply by innermatrix** [reply]
Thanks. I’ll revisit that when I can get off 0.20 🙂

**Reply by FredKSchott** [reply]
np, thanks for the feedback!

**Comment by matthewp** [comment]
I think it would be good to include in the proposal a grouping of use-cases and how those would be achieved. Some examples:

1. Including an external stylesheet that gets built.
2. Including an external stylesheet that doesn't get built.
3. Including scoped styles
4. Including styles in a JSX component, etc.

**Reply by FredKSchott** [reply]
Agreed, these should be more explicitly highlighted in the rough proposal.

RE 4.: I'm hoping to keep this out-of-scope, and not propose any changes to how CSS is currently supported outside of Astro components and only focus on Astro components.

**Comment by matthewp** [comment]
Assuming the `?url` feature worked as we expected what should happen here?

```astro
---
import one from './styles/one.css?url'
import two from './styles/two.css?url'
---

<link rel="stylesheet" href={one}>
<link rel="stylesheet" href={two}>
```

1. Should these stylesheets be bundled together?
2. If they were bundled together there will be 2 links to the same URL. Is that ok?
3. If they were bundled it would likely be based on import order and not link-usage order. This could result in unexpected specificity.

For the above reasons I think it makes more sense that, if we support this import syntax, they not be bundled. And instead treated like image assets where it is hashed and moved into `dist/`.

**Reply by matthewp** [reply]
Also consider tools like this: https://github.com/aFuzzyBear/astro-academy/tree/master/packages/astro-ui/head/stylesheet

We want to enable these things to continue to exist, and this shows a use-case where you set specific media attributes like `<link rel="stylesheet media="print">` which should *not* be bundled.

**Reply by FredKSchott** [reply]
Based on the bug in Vite (that I can't image certainly won't be fixed until v3.0) it may be better to just explicitly disable support for this for now. If Vite won't support `import two from './styles/two.css?url'` until v3.0, then I don't think it's worth us going too deep on this, and whether Vite would bundle or not bundle those two CSS files correctly for us. The other options in the proposal handle this use-case.

But yes, to be super clear, link would not bundle assets. Went into more detail below: https://github.com/withastro/rfcs/discussions/28#discussioncomment-1740501

**Reply by sebiko3** [reply]
Wouldn't it be possible to add some attribute to link tags which flag them as "not to be bundled"? 

```js
---
import one from './styles/one.css?url'
import two from './styles/two.css?url'
import three from './styles/three.css?url'
---
<link rel="stylesheet" href={one}>
<link rel="stylesheet" href={two} nobundle>
<link rel="stylesheet" href={three} nobundle>
```

**Reply by FredKSchott** [reply]
The rough proposal above is making the point that `<link>` would not bundle by default. 

We thought that `import one from './styles/one.css?url'` did create a bundled CSS file in your final output, in which case the two could be used together. However, the Vite bug will most likely block this, in which case the `<link>` tag would only support references to files in `public/`, without any bundling.

**Reply by sebiko3** [reply]
ah gotcha! okay that would do as well, for me at least.

**Reply by matthewp** [reply]
@FredKSchott is __(3)__ the one that handles this use case? If so, would the external CSS be scoped? This use case is for bundling external CSS without scoping it, for clarity.

**Reply by FredKSchott** [reply]
Ah, that should definitely be clarified. I don't believe that a CSS `@import` is currently scoped to the component. If that's undefined then I can make the RFC make it explicit.
